My Project/clsTranslateWinForms.vb

```diff
@@ -1 +1,637 @@
+                                                   strLanguage As String) As String
+        'Create a list of all the tool tip objects associated with this (sub)dialog
+        'Note: Normally, a (sub)dialog wil only have a single tool tip object. This stores the 
+        '      tool tips for all the components in the (sub)dialog.
+        Dim lstToolTips = New List(Of ToolTip)
+        For Each clsDctEntry As KeyValuePair(Of String, Component) In dctComponents
+            'TODO for efficiency, we assume that only forms and user controls have tool tip objects.
+            'Also allow other component types to have tool tip objects?
+            If Not (TypeOf clsDctEntry.Value Is Form) AndAlso Not (TypeOf clsDctEntry.Value Is UserControl) Then
+                Continue For
+            End If
+            'Tool tip objects are stored in the form's private list of components. There is no 
+            '    public access. Therefore we need multiple steps to get the list of tool tip objects.
+            Dim clsType As Type = clsDctEntry.Value.GetType()
+            Dim clsFieldInfo As FieldInfo = clsType?.GetField("components", BindingFlags.Instance Or BindingFlags.NonPublic)
+            Dim clsContainer As Container = clsFieldInfo?.GetValue(clsDctEntry.Value)
+            Dim lstToolTipsCtrl As List(Of ToolTip) = clsContainer?.Components.OfType(Of ToolTip).ToList()
+            If lstToolTipsCtrl IsNot Nothing Then
+                For Each clsToolTip As ToolTip In lstToolTipsCtrl
+                    lstToolTips.Add(clsToolTip)
+                Next
+            End If
+        Next
+
+        Try
+            'connect to the SQLite database that contains the translations
+            Dim clsBuilder As New SQLiteConnectionStringBuilder With {
+                .FailIfMissing = True,
+                .DataSource = strDataSource}
+            Using clsConnection As New SQLiteConnection(clsBuilder.ConnectionString)
+                clsConnection.Open()
+                Using clsCommand As New SQLiteCommand(clsConnection)
+
+                    'get all static translations for the specified form and language
+                    clsCommand.CommandText =
+                            "SELECT control_name, form_controls.id_text, translation " &
+                            "FROM form_controls, translations WHERE form_name = '" & strControlName &
+                            "' AND language_code = '" & strLanguage &
+                            "' AND form_controls.id_text = translations.id_text"
+                    Dim clsReader As SQLiteDataReader = clsCommand.ExecuteReader()
+                    Using clsReader
+
+                        'for each translation row
+                        While (clsReader.Read())
+
+                            'ignore rows where the translation text is null or missing
+                            If clsReader.FieldCount < 3 OrElse clsReader.IsDBNull(2) Then
+                                Continue While
```

Comment:
@lloyddewit in line 402, I'm struggling to find a scenario where the `id_text` will be missing or null. I think that should never happen and the check should also be enforced at the database level.

And if there are such scenarios, shouldn't we just add this condition as part of the `sql` text in line 390? I usually find that to be easier and much more efficient, that is, only if its necessary to have that check.

Reply:
I agree. I removed the check and on line 37, I specified that columns should not be null:

```
''' This class uses an SQLite database to translate text items to a new language. The database must 
''' contain the following tables:
''' <code>
''' CREATE TABLE "form_controls" (
'''	"form_name"	TEXT,
'''	"control_name"	TEXT,
'''	"id_text"	TEXT NOT NULL,
'''	PRIMARY KEY("form_name", "control_name")
''' )
''' <para>
''' CREATE TABLE "translations" (
'''	"id_text"	TEXT,
'''	"language_code"	TEXT,
'''	"translation"	TEXT NOT NULL,
'''	PRIMARY KEY("id_text", "language_code")
''' )
''' </para></code>
```

Reply:
I removed the null checks and the field count checks.